src/components/PostDetails.tsx

```diff
@@ -1,118 +1,148 @@
-import React from 'react';
+import React, { useCallback, useEffect, useState } from 'react';
+import { addComment, deleteComment, getComments } from '../api/comments';
+import { Comment } from '../types/Comment';
+import { Errors } from '../types/Errors';
+import { Post } from '../types/Post';
+import { CommentItem } from './CommentItem';
+import { ErrorNotification } from './ErrorNotification';
 import { Loader } from './Loader';
 import { NewCommentForm } from './NewCommentForm';
 
-export const PostDetails: React.FC = () => {
-  return (
-    <div className="content" data-cy="PostDetails">
-      <div className="content" data-cy="PostDetails">
-        <div className="block">
-          <h2 data-cy="PostTitle">
-            #18: voluptate et itaque vero tempora molestiae
-          </h2>
-
-          <p data-cy="PostBody">
-            eveniet quo quis
-            laborum totam consequatur non dolor
-            ut et est repudiandae
-            est voluptatem vel debitis et magnam
-          </p>
-        </div>
+type Props = {
+  postDetails: Post;
+  isNewCommentFormOpened: boolean;
+  onOpenNewCommentForm: () => void;
+};
+
+export const PostDetails: React.FC<Props> = React.memo(
+  ({
+    postDetails,
+    isNewCommentFormOpened,
+    onOpenNewCommentForm,
+  }) => {
+    const {
+      id,
+      title,
+      body,
+    } = postDetails;
+
+    const [comments, setComments] = useState<Comment[]>([]);
+    const [isLoading, setIsLoading] = useState(false);
+    const [isAddingComment, setIsAddingComment] = useState(false);
+    const [isCommentsLoaded, setIsCommentsLoaded] = useState(false);
+    const [isCommentsLoadingError, setIsCommentsLoadingError] = useState(false);
+
+    useEffect(() => {
+      setIsLoading(true);
+      setComments([]);
+
+      getComments(id)
+        .then(loadedComments => {
+          setComments(loadedComments);
+          setIsCommentsLoaded(true);
+          setIsCommentsLoadingError(false);
+        })
+        .catch(() => {
+          setIsCommentsLoadingError(true);
+        })
+        .finally(() => setIsLoading(false));
+    }, [id]);
+
+    const addNewComment = (comment: Omit<Comment, 'id'>) => {
+      setIsAddingComment(true);
 
-        <div className="block">
-          <Loader />
+      addComment(comment)
+        .then(addedComment => (
+          setComments(prevComments => [...prevComments, addedComment])
+        ))
+        .finally(() => setIsAddingComment(false));
+    };
+
+    const deleteSelectedComment = useCallback((commentId: number) => {
+      deleteComment(commentId);
+
+      setComments(prevComments => prevComments.filter(comment => (
+        comment.id !== commentId
+      )));
+    }, []);
+
+    const isLoadedCommentsExist = !isLoading
+      && !isCommentsLoadingError
+      && isCommentsLoaded
+      && comments.length > 0;
+
+    const isLoadedCommentsNotExist = !isLoading
+      && !isCommentsLoadingError
+      && isCommentsLoaded
+      && !comments.length;
```

Comment:
it's not critical, but do you really need two of these variables?

Reply:
I think it is necessary to have two variables in order to display correctly the loader / error / desired message / list. It makes the code bigger but more convenient UI for user